libbeat/processors/decode_xml_wineventlog/processor.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/elastic/beats/v7/winlogbeat/sys/winevent"
+)
+
+var (
+	errFieldIsNotString = errors.New("field value is not a string")
+)
+
+const (
+	procName = "decode_xml_wineventlog"
+	logName  = "processor." + procName
+)
+
+func init() {
+	processors.RegisterPlugin(procName,
+		checks.ConfigChecked(New,
+			checks.RequireFields("field"),
+			checks.AllowedFields(
+				"field", "target_field",
+				"overwrite_keys", "document_id",
+				"ignore_missing", "ignore_failure",
+			)))
+	jsprocessor.RegisterPlugin(procName, New)
+}
+
+type processor struct {
+	config
+
+	decoder decoder
+	log     *logp.Logger
+}
+
+type decoder interface {
+	decode(data []byte) (win, ecs common.MapStr, err error)
+}
+
+// New constructs a new decode_xml processor.
+func New(c *common.Config) (processors.Processor, error) {
+	config := defaultConfig()
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("fail to unpack the "+procName+" processor configuration: %s", err)
+	}
+
+	return newProcessor(config)
+}
+
+func newProcessor(config config) (processors.Processor, error) {
+	cfgwarn.Experimental("The " + procName + " processor is experimental.")
+
+	return &processor{
+		config:  config,
+		decoder: newDecoder(),
+		log:     logp.NewLogger(logName),
+	}, nil
+}
+
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	if err := p.run(event); err != nil && !p.IgnoreFailure {
+		err = fmt.Errorf("failed in decode_xml on the %q field: %w", p.Field, err)
+		event.PutValue("error.message", err.Error())
+		return event, err
+	}
+	return event, nil
+}
+
+func (p *processor) run(event *beat.Event) error {
+	data, err := event.GetValue(p.Field)
+	if err != nil {
+		if p.IgnoreMissing && err == common.ErrKeyNotFound {
+			return nil
+		}
+		return err
+	}
+
+	text, ok := data.(string)
+	if !ok {
+		return errFieldIsNotString
+	}
+
+	win, ecs, err := p.decoder.decode([]byte(text))
+	if err != nil {
+		return fmt.Errorf("error decoding XML field: %w", err)
+	}
+
+	var id string
+	if tmp, err := common.MapStr(win).GetValue(p.DocumentID); err == nil {
+		if v, ok := tmp.(string); ok {
+			id = v
+			common.MapStr(win).Delete(p.DocumentID)
+		}
+	}
+
+	if p.Target != "" {
+		if _, err = event.PutValue(p.Target, win); err != nil {
+			return fmt.Errorf("failed to put value %v into field %q: %w", win, p.Target, err)
+		}
+	} else {
+		jsontransform.WriteJSONKeys(event, win, false, p.OverwriteKeys, !p.IgnoreFailure)
+	}
+
+	if p.MapECSFields {
+		jsontransform.WriteJSONKeys(event, ecs, false, p.OverwriteKeys, !p.IgnoreFailure)
+	}
+
+	if id != "" {
+		event.SetID(id)
+	}
+
+	return nil
+}
+
+func (p *processor) String() string {
+	json, _ := json.Marshal(p.config)
+	return procName + "=" + string(json)
+}
+
+func fields(evt winevent.Event) (common.MapStr, common.MapStr) {
+	win := evt.Fields()
+
+	ecs := common.MapStr{}
+
+	ecs.Put("event.kind", "event")
+	ecs.Put("event.code", fmt.Sprint(evt.EventIdentifier.ID))
```

Comment:
even if it is numeric in winlog, I am not sure if we should put it as such in event.code, since it is a keyword according to https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-code

Reply:
I guess the question is if we want to change winlog (I'm for it).  The problem is doing comparisons on `event.code`, it gets to be a real pain if we have to check it as both a number and a string.  And the pipeline needs to work for both winlog & decode_xml_winlogevent.

Reply:
Maybe is a good idea to keep it as a number here and do the type change all at once in its own PR then?